micropython/test_micropython.c

```diff
@@ -0,0 +1,126 @@
+#ifdef __STDC_ALLOC_LIB__
+#define __STDC_WANT_LIB_EXT2__ 1
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+
+#define MICROPYTHON   "/bin/micropython "
+#define PATH_TO_TESTS "/usr/test/micropython/"
+
+#define CMDLINE_OPTION_IDEN "# cmdline: "
+
+void exit_with_error(char *msg)
+{
+	fprintf(stderr, "Error: %s\n", msg);
+	exit(1);
+}
+
+char *cat(const char *str1, const char *str2)
+{
+	int length = strlen(str1) + strlen(str2) + 1;
+	char *result = (char *)malloc(sizeof(char) * length);
+	if (!result) {
+		return NULL;
+	}
+
+	result[0] = '\0';
+	strcat(result, str1);
+	strcat(result, str2);
+
+	return result;
+}
+
+char *check_for_options(const char *path)
+{
+	FILE *f;
+	char *line = NULL;
+	size_t len, result_len, opt_iden_len = strlen(CMDLINE_OPTION_IDEN);
+	ssize_t read;
+	char *result;
+
+	f = fopen(path, "r");
+	if (!f)
+		exit_with_error("There was a problem with opening file with a script");
+
+	if (getline(&line, &len, f) < 0)
+		exit_with_error("There was a problem with reading script to execute");
+
+	len = strlen(line) + 1;
+
+	if (len > opt_iden_len &&
+		strncmp(line, CMDLINE_OPTION_IDEN, opt_iden_len) == 0) {
+		result_len = strlen(line) - opt_iden_len;
+		result = (char *)malloc(sizeof(char) * result_len);
+		if (!result)
+			exit_with_error("There is no enough memory to run this test");
+
+		strcpy(result, &line[opt_iden_len]);
+
+		/* Changing the last character from '\n' to ' ' */
+		result[strlen(result) - 1] = ' ';
+	}
+	else
+		result = NULL;
+
+	free(line);
+
+	return result;
+}
+
+int main(int argc, char **argv)
+{
+	char *path, *cmd, *options, *tmp;
+	int t;
+
+	if (argc != 2)
+		exit_with_error("There is no required argument - please execute this function with path to the micropython tests");
+
+	if (chdir(PATH_TO_TESTS) != 0)
+		exit_with_error("There is no such a micropython test to run, build project with \"LONG_TEST=y\"");
+
+	path = cat(PATH_TO_TESTS, argv[1]);
+	if (!path)
+		exit_with_error("There is no enough memory to run this test");
+
+	printf("Running test: %s\n", path);
+	free(path);
+
+	if (access(argv[1], F_OK) != 0)
+		exit_with_error("There is no such a micropython test to run, build project with \"LONG_TEST=y\"");
+
+	options = check_for_options(argv[1]);
+	if (options) {
+		tmp = cat(MICROPYTHON, options);
+		if (!tmp)
+			exit_with_error("There is no enough memory to run this test");
+
+		cmd = cat(tmp, argv[1]);
+		if (!cmd) {
+			free(tmp);
+			exit_with_error("There is no enough memory to run this test");
+		}
+
+		free(tmp);
+	}
+	else {
+		cmd = cat(MICROPYTHON, argv[1]);
+		if (!cmd)
+			exit_with_error("There is no enough memory to run this test");
+	}
+
+	if ((t = system(cmd)) == 1) {
+		free(cmd);
+		exit_with_error("There was an error during execution micropython test");
+	}
+	else if (t) {
+		free(cmd);
+		exit_with_error("There was an error caused by function system() (not micropython test)");
+	}
```

Comment:
Maybe here we can drop the `exit_with_error` function and just use `printf()`. No matter if an error occurs we call `free(cmd)`, so that just adds boilerplate code.